### ISSUE TITLE
1000x Faster Data Augmentation  The Berkeley Artificial Intelligence Research Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5534]
1000x Faster Data Augmentation &ndash; The Berkeley Artificial Intelligence Research Blog<br>
<br>
https://ift.tt/2Lqjk9b<br>
<br>
Daniel Seita<br>
<br>
<div>
<article>
<p><img src="https://bair.berkeley.edu/static/blog/data_aug/augmentation_schedule_viz.png"><br>
<i>Effect of Population Based Augmentation applied to images, which differs at different percentages into training.</i></p>
<p>In this blog post we introduce Population Based Augmentation (PBA), an algorithm that quickly and efficiently learns a state-of-the-art approach to augmenting data for neural network training. PBA matches the previous best result on CIFAR and SVHN but uses <b><i>one thousand times less compute</i></b>, enabling researchers and practitioners to effectively learn new augmentation policies using a single workstation GPU. You can use PBA broadly to improve deep learning performance on image recognition tasks.</p>
<p>We discuss the PBA results from our <a href="https://arxiv.org/abs/1905.05393.pdf">recent paper</a> and then show how to easily <a href="https://github.com/arcelien/pba">run PBA for yourself</a> on a new data set in the <a href="https://ray.readthedocs.io/en/latest/tune.html">Tune</a> framework.</p>
<p>Recent advances in deep learning models have been largely attributed to the quantity and diversity of data gathered in recent years. Data augmentation is a strategy that enables practitioners to significantly increase the diversity of data available for training models, without actually collecting new data. Data augmentation techniques such as cropping, padding, and horizontal flipping are commonly used to train large neural networks. However, most approaches used in training neural networks only use basic types of augmentation. While neural network architectures have been investigated in depth, less focus has been put into discovering strong types of data augmentation and data augmentation policies that capture data invariances.</p>
<p><img src="https://bair.berkeley.edu/static/blog/data_aug/basic_aug.png"><br>
<i>An image of the number &ldquo;3&rdquo; in original form and with basic augmentations applied.</i></p>
<p>Recently, Google has been able to push the state-of-the-art accuracy on datasets such as CIFAR-10 with <a href="https://arxiv.org/abs/1805.09501">AutoAugment</a>, a new automated data augmentation technique. AutoAugment has shown that prior work using just applying a fixed set of transformations like horizontal flipping or padding and cropping leaves potential performance on the table. AutoAugment introduces 16 geometric and color-based transformations, and formulates an augmentation <em>policy</em> that selects up to two transformations at certain magnitude levels to apply to each batch of data. These higher performing augmentation policies are learned by training models directly on the data using reinforcement learning.</p>
<h2>What&rsquo;s the catch?</h2>
<p>AutoAugment is a very expensive algorithm which requires training 15,000 models to convergence to generate enough samples for a reinforcement learning based policy. No computation is shared between samples, and it costs 15,000 NVIDIA Tesla P100 GPU hours to learn an ImageNet augmentation policy and 5,000 GPU hours to learn an CIFAR-10 one. For example, if using Google Cloud on-demand P100 GPUs, it would cost about <span>$</span>7,500 to discover a CIFAR policy, and <span>$</span>37,500 to discover an ImageNet one! Therefore, a more common use case when training on a new dataset would be to transfer a pre-existing published policy, which the authors show works relatively well.</p>
<p>Our formulation of data augmentation policy search, Population Based Augmentation (PBA), reaches similar levels of test accuracy on a variety of neural network models while utilizing three orders of magnitude less compute. We learn an augmentation policy by training several copies of a small model on CIFAR-10 data, which takes five hours using a NVIDIA Titan XP GPU. This policy exhibits strong performance when used for training from scratch on larger model architectures and with CIFAR-100 data.</p>
<p>Relative to the several days it takes to train large CIFAR-10 networks to convergence, the cost of running PBA beforehand is marginal and significantly enhances results. For example, training a PyramidNet model on CIFAR-10 takes over 7 days on a NVIDIA V100 GPU, so learning a PBA policy adds only 2% precompute training time overhead. This overhead would be even lower, under 1%, for SVHN.</p>
<p><img src="https://bair.berkeley.edu/static/blog/data_aug/chart.png"><br>
<i>CIFAR-10 test set error between PBA, AutoAugment, and the baseline which only uses horizontal flipping, padding, and cropping, on <a href="https://arxiv.org/abs/1605.07146">WideResNet</a>, <a href="https://arxiv.org/abs/1705.07485">Shake-Shake</a>, and <a href="https://arxiv.org/abs/1610.02915">PyramidNet</a>+<a href="https://arxiv.org/abs/1802.02375">ShakeDrop</a> models. PBA is significantly better than the baseline and on-par with AutoAugment.</i></p>
<p>PBA leverages the <a href="https://deepmind.com/blog/population-based-training-neural-networks/">Population Based Training algorithm</a> to generate an augmentation policy <em>schedule</em> which can adapt based on the current epoch of training. This is in contrast to a fixed augmentation policy that applies the same transformations independent of the current epoch number.</p>
<p>This allows an ordinary workstation user to easily experiment with the search algorithm and augmentation operations. One interesting use case would be to introduce new augmentation operations, perhaps targeted towards a particular dataset or image modality, and be able to quickly produce a tailored, high performing augmentation schedule. Through ablation studies, we have found that the learned hyperparameters and schedule order are important for good results.</p>
<h2>How is the augmentation schedule learned?</h2>
<p>We use Population Based Training with a population of 16 small WideResNet models. Each worker in the population will learn a different candidate hyperparameter schedule. We transfer the best performing schedule to train larger models from scratch, from which we derive our test error metrics.</p>
<p><img width="600" src="https://bair.berkeley.edu/static/blog/data_aug/pbt_visual.png"><br>
<i>Overview of Population Based Training, which discovers hyperparameter schedules by training a population of neural networks. It combines random search (explore) with the copying of model weights from high performing workers (exploit). <a href="https://deepmind.com/blog/population-based-training-neural-networks/">Source</a></i></p>
<p>The population models are trained on the target dataset of interest starting with all augmentation hyperparameters set to 0 (no augmentations applied). At frequent intervals, an &ldquo;exploit-and-explore&rdquo; process &ldquo;exploits&rdquo; high performing workers by copying their model weights to low performing workers, and then &ldquo;explores&rdquo; by perturbing the hyperparameters of the worker. Through this process, we are able to share compute heavily between the workers and target different augmentation hyperparameters at different regions of training. Thus, PBA is able to avoid the cost of training thousands of models to convergence in order to reach high performance.</p>
<p>We leverage Tune&rsquo;s built-in implementation of PBT to make it straightforward to use PBA.</p>
<div>
<div>
<pre><code><span>import</span> <span>ray</span>
<span>def</span> <span>explore</span><span>(</span><span>config</span><span>):</span>
    <span>"""Custom PBA function to perturb augmentation hyperparameters."""</span>
    <span>...</span>

<span>ray</span><span>.</span><span>init</span><span>()</span>
<span>pbt</span> <span>=</span> <span>ray</span><span>.</span><span>tune</span><span>.</span><span>schedulers</span><span>.</span><span>PopulationBasedTraining</span><span>(</span>
    <span>time_attr</span><span>=</span><span>"training_iteration"</span><span>,</span>
    <span>reward_attr</span><span>=</span><span>"val_acc"</span><span>,</span>
    <span>perturbation_interval</span><span>=</span><span>3</span><span>,</span>
    <span>custom_explore_fn</span><span>=</span><span>explore</span><span>)</span>
<span>train_spec</span> <span>=</span> <span>{</span><span>...</span><span>}</span>  <span># Things like file paths, model func, compute.</span>
<span>ray</span><span>.</span><span>tune</span><span>.</span><span>run_experiments</span><span>({</span><span>"PBA"</span><span>:</span> <span>train_spec</span><span>},</span> <span>scheduler</span><span>=</span><span>pbt</span><span>)</span>
</code></pre>
</div>
</div>
<p>We call Tune&rsquo;s implementation of PBT with our custom exploration function. This will create 16 copies of our WideResNet model and train them time-multiplexed. The policy schedule used by each copy is saved to disk and can be retrieved after termination to use for training new models.</p>
<p>You can run PBA by following the README at: <a href="https://github.com/arcelien/pba">https://github.com/arcelien/pba</a>. On a Titan XP, it only requires one hour to learn a high performing augmentation policy schedule on the SVHN dataset. It is also easy to use PBA on a custom dataset as well: simply define a new dataloader and everything else falls into place.</p>
<p>Big thanks to Daniel Rothchild, Ashwinee Panda, Aniruddha Nrusimha, Daniel Seita, Joseph Gonzalez, and Ion Stoica for helpful feedback while writing this post. Feel free to get in touch with us on <a href="https://github.com/arcelien/pba">Github</a>!</p>
<p>This post is based on the following paper to appear in ICML 2019 as an oral presentation:</p>
<ul>
<li>
<b>Population Based Augmentation: Efficient Learning of Augmentation Policy Schedules</b><br>
Daniel Ho, Eric Liang, Ion Stoica, Pieter Abbeel, Xi Chen<br>
<a href="https://arxiv.org/abs/1905.05393">Paper</a> <a href="https://github.com/arcelien/pba">Code</a>
</li>
</ul>
</article>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via The Berkeley Artificial Intelligence Research Blog https://ift.tt/2EPVVfk<br>
<br>
September 27, 2019 at 11:55AM
